src/apps/MASS3DPA-Seq.cpp

```diff
@@ -0,0 +1,195 @@
+
+#include "MASS3DPA.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#include <iostream>
+
+namespace rajaperf {
+namespace apps {
+
+#define RAJA_UNROLL(N)
+#define FOREACH_THREAD(i, k, N) for (int i = 0; i < N; i++)
+
+void MASS3DPA::runSeqVariant(VariantID vid) {
+  const Index_type run_reps = getRunReps();
+
+  MASS3DPA_DATA_SETUP;
+
+  switch (vid) {
+
+  case Base_Seq: {
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      for (int e = 0; e < NE; ++e) {
+
+        MASS3DPA_0_CPU
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(dx, x, D1D){MASS3DPA_1}
+          FOREACH_THREAD(dx, x, Q1D) {MASS3DPA_2}
+        }
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(qx, x, Q1D) { MASS3DPA_3 }
+        }
+
+        FOREACH_THREAD(qy, y, Q1D) {
+          FOREACH_THREAD(qx, x, Q1D) { MASS3DPA_4 }
+        }
+
+        FOREACH_THREAD(qy, y, Q1D) {
+          FOREACH_THREAD(qx, x, Q1D) { MASS3DPA_5 }
+        }
+
+        FOREACH_THREAD(d, y, D1D) {
+          FOREACH_THREAD(q, x, Q1D) { MASS3DPA_6 }
+        }
+
+        FOREACH_THREAD(qy, y, Q1D) {
+          FOREACH_THREAD(dx, x, D1D) { MASS3DPA_7 }
+        }
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(dx, x, D1D) { MASS3DPA_8 }
+        }
+
+        FOREACH_THREAD(dy, y, D1D) {
+          FOREACH_THREAD(dx, x, D1D) { MASS3DPA_9 }
+        }
+
+      } // element loop
+    }
+    stopTimer();
+
+    break;
+  }
+
+#if defined(RUN_RAJA_SEQ)
+  case RAJA_Seq: {
+
+    //Currently Teams requires two policies if compiled with a device
+    using launch_policy = RAJA::expt::LaunchPolicy<RAJA::expt::seq_launch_t
+#if defined(RAJA_DEVICE_ACTIVE)
+                                                   ,device_launch
+#endif
+                                                   >;
+
+    using teams_x = RAJA::expt::LoopPolicy<RAJA::loop_exec
+#if defined(RAJA_DEVICE_ACTIVE)
+                                           ,gpu_block_x_policy
+#endif
+                                           >;
+
+    using threads_x = RAJA::expt::LoopPolicy<RAJA::loop_exec
+#if defined(RAJA_DEVICE_ACTIVE)
+                                             ,gpu_thread_x_policy
+#endif
+                                             >;
+
+    using threads_y = RAJA::expt::LoopPolicy<RAJA::loop_exec
+#if defined(RAJA_DEVICE_ACTIVE)
+                                             ,gpu_thread_y_policy
+#endif
+                                             >;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      RAJA::expt::launch<launch_policy>(
+        RAJA::expt::HOST, RAJA::expt::Resources(),
+          [=] RAJA_HOST_DEVICE(RAJA::expt::LaunchContext ctx) {
+            RAJA::expt::loop<teams_x>(ctx, RAJA::RangeSegment(0, NE), [&](int e) {
+
+                  MASS3DPA_0_CPU
+
+                  RAJA::expt::loop<threads_y>(ctx, RAJA::RangeSegment(0, D1D), [&](int dy) {
```

Comment:
I'm also confused by the "threads" terminology in this sequential file. Would it be clearer to name the policies to better align with how the kernel variants in the files will execute?

Reply:
That is a good point; In general I was going for an abstraction of a hierarchical nested loop model that "fits" the block/thread model of GPU computing. I suppose we can refer the policies as "inner_i/outer_i" policies like that they are universal across programming models .